py-rse/style.Rmd

```diff
@@ -287,755 +349,875 @@ n = count(lines)
+or put them in the wrong order.
+One solution is to bundle parameters together so that (for example)
+people pass three `point` objects instead of nine separate `x`, `y`, and `z` values.
+
+A second approach (which can be combined with the previous one)
+is to specify default values for some of the parameters.
+Doing this gives users control over everything
+while also allowing them to ignore details;
+it also codifies what you consider "normal" for the function.
+
+For example,
+suppose we are comparing images to see if they are the same or different.
+We can specify two kinds of tolerance:
+how large a difference in color value to notice,
+and how many differences above that threshold to tolerate
+(as a percentage of the total number of pixels).
+By default,
+any color difference is considered significant,
+and only 1% of pixels are allowed to differ:
 
 ```python
-# ...after...
-def rescale_by_average(values, factors, weights):
-    if (not values) or (not factors) or (not weights):
-        return
-    a = 0.0
-    for (f, w) in zip(factors, weights):
-        a += f * w
-    a /= len(f)
-    for (i, v) in enumerate(values):
-        values[i] = v / a
+def image_diff(left, right, per_pixel=0, fraction=0.01):
+    # ...implementation...
 ```
 
-A related refactoring pattern is called "default and override".
-To use it,
-find cases where a value is set conditionally;
-assign the default or most common value unconditionally,
-and then override it in a special case.
-The result is fewer lines of code and clearer control flow;
-however,
-it does mean executing two assignments instead of one,
-so it shouldn't be used if the common case is expensive
-(e.g., involves a database lookup or a web request).
+When this function is called using `image_diff(old, new)`,
+those default values apply.
+However,
+it can also be called like this:
+
+-   `image_diff(old, new, per_pixel=2)`
+    allows pixels to differ slightly without those differences being significant.
+-   `image_diff(old, new, fraction=0.05)` allows more pixels to differ.
+-   `image_diff(old, new, per_pixel=1, fraction=0.005)`
+    raises the per-pixel threshold but decrease number of allowed differences.
 
-```python
-# ...before..
-if configuration['threshold'] > UPPER_BOUND:
-    scale = 0.8
-else:
-    scale = 1.0
-```
+Default parameter values make code easier to understand and use,
+but there is a subtle trap.
+When Python executes a function definition like this:
 
 ```python
-# ...after...
-scale = 1.0
-if configuration['threshold'] > UPPER_BOUND:
-    scale = 0.8
+def collect(new_value, accumulator=set()):
+    accumulator.add(new_value)
+    return accumulator
 ```
 
-In simple cases,
-people will sometimes put the test and assignment on a single line:
+it calls `set()` to create a new empty set,
+and then uses that set as the default value for `accumulator` every time the function is called.
+It does *not* call `set()` once for each call,
```

Comment:
Mention that this is called "mutable defaults" and link either the docs (https://docs.python.org/3/library/dataclasses.html#mutable-default-values was the best I could find) or something clearer like http://effbot.org/zone/default-values.htm

Reply:
Does that jargon add value for our readers?

Reply:
I think the added value is that the name and link are useful for leaning more, and that the mention of the term here where we clearly explain it makes it less intimidating to come across it later (where it might not have a good explanation).

Reply:
I'd never actually run into the phrase "mutable defaults" before you mentioned it - how do others feel?